### ISSUE TITLE
Help Stefano to fix Content-Length retrieval feature in Daniil's client

### ISSUE THREAD

**Issue by 552020** [dantol29/webserver#143]
I'm working on the branch ‹slombard/web-113-fix-poll-client-socket-error-bad-file-descriptor›. I refactored a little bit Daniil's (@dtolmaco) client, but I'm failing at one point and I don't understand why and this is slowly driving me crazy. <br><br>The feature I tried to add to Daniil's client:

<br>Before sending the request (that we pass as an argument to the ./client executable) to the server the client should check if the declared Content-Lenght of the body in the header correspond to the size of the header. And I'm failing at this check that should be easy. I wanted to add this, because if the Content-Length value is not equal to length of the body the server will close the connection, this being the correct behaviour of the server.<br><br>The issue:<br>In the extractContentLenght (I have two similar in the code `extractContentLength` and `extractContentLengthAlt`) function(s) we try first to individuate the Content-Length header, then we try to read the value of the this header. The syntax of this Header should be something like  like `Content-Length: 86\n\r`. So we search for `Content-Length:` in the http request, then we skip from that position by 15 or 16 chars (two different approaches in `extractContentLength` and `extractContentLengthAlt`) of `Content-Length` depending we count the space between the key and th vaue or not and then we look for the next \\n\\r, to know where the end of the number defining the Content-Length is, and we try to put `endPos` to that value with `find`. This passage is failing, my function is unable to find `\n\r` and therefore it behave as it would not be a CLRF after the Content-ength header and it returns `Malformed HTTP header, no CRLF after Content-Length.`

<img src="https://uploads.linear.app/3a13e0cf-cf78-41bb-abe3-8d0765670e8b/b92cda2e-0c00-4da4-9aba-d24c9de2bb4a/be397e52-8244-435c-9ac2-6a97c63dc45b?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzNhMTNlMGNmLWNmNzgtNDFiYi1hYmUzLThkMDc2NTY3MGU4Yi9iOTJjZGEyZS0wYzAwLTRkYTQtOWFiYS1kMjRjOWRlMmJiNGEvYmUzOTdlNTItODI0NC00MzVjLTlhYzItNmE5N2M2M2RjNDViIiwiaWF0IjoxNzEzMTY3NTgxLCJleHAiOjMzMjgzNzI3NTgxfQ.Ikdy85dnV7e66rkbbwvVQWp2tXYwQyIBS9CGQ_dsiMQ " alt="image.png" width="573" height="421" />

The request I'm testing it with is:

```
./client $'POST / HTTP/1.1\r\nHost: www.example.com\r\nContent-Type: multipart/form-data; boundary=stop_please\r\nContent-Length: 97\r\n\r\n--stop_please\r\nContent-Disposition: form-data; name="field1"\r\n\r\nblablabla\r\n--stop_please--\r\n'
```

I also print it inside the function to check if it is imported properly from the terminal into the client program (and it is the case) but still I'm not able to find the bug. HELP!<br><br>I assign this to @dtolmaco cause probably is the one that understand better his own client program, but for sure also @lmangall can have a look at it. <br><br>It's not a super necessary feature of the test client but it should be also super easy to implement and I don't understand why I'm failing. <br><br>Otherwise the code in this branch should work, i.e. should solve the body parsing issue I was working on this branch.

**Comment by dantol29** [comment]
First of all, why is there a double CRLF after content length?
Second of all, I think that 'core' should not parse anything from the header, it doesn't make sense. It is already implemented in the parser  and you can feed it even before you read body.

**Comment by 552020** [comment]
> First of all, why is there a double CRLF after content length?

Cause the header is to the end and the body starts, is this not correct?

> Second of all, I think that 'core' should not parse anything from the header, it doesn't make sense. It is already implemented in the parser and you can feed it even before you read body.

Yes, it needs to do so. Cause it needs to know if the request is chunked or not. And if it's not chunked he needs to know how much it has to read from the client, to assure the the whole content arrived. This also to prevent certain type of attacks where you send an infinite long header basically overwhelming the server, so the header as a max length, that should be checkd by the server and it is the same for the body if the request is not chunked.

**Comment by 552020** [comment]
Probably it is also possible also to have a core that just blindly read from the socket, but then after each read it should pass the content to the parser that analyze the content and tell what to do next. I gues it would be more complicated for us cause it should be a multiple back and forth between core and parser.

**Comment by dantol29** [comment]
You can read until you encounter \\r\\n\\r\\n (the end of the header) and then pass it to the parser

**Comment by 552020** [comment]
Yes I know. I opened a linear issue about it. We can discuss it there, cause it would mean a back and forth between core and parser.

**Comment by dantol29** [comment]
We can close it since it is not responsibility of the tester to check content length.
